Sources/SwiftGRPCNIO/GRPCHTTPProtocolSwitcher.swift

```diff
@@ -0,0 +1,97 @@
+import Foundation
+import NIO
+import NIOHTTP1
+import NIOHTTP2
+
+/// Channel handler that creates different processing pipelines depending on whether
+/// the incoming request is HTTP 1 or 2.
+class GRPCHTTPProtocolSwitcher {
+  private let handlersInitializer: ((Channel) -> EventLoopFuture<Void>)
+
+  public init(handlersInitializer: (@escaping (Channel) -> EventLoopFuture<Void>)) {
+    self.handlersInitializer = handlersInitializer
+  }
+}
+
+extension GRPCHTTPProtocolSwitcher: ChannelInboundHandler {
+  typealias InboundIn = ByteBuffer
+  typealias InboundOut = ByteBuffer
+
+  enum HTTPProtocolVersionError: Error {
+    /// Raised when it wasn't possible to detect HTTP Protocol version.
+    case invalidHTTPProtocolVersion
+
+    var localizedDescription: String {
+      switch self {
+      case .invalidHTTPProtocolVersion:
+        return "Could not identify HTTP Protocol Version"
+      }
+    }
+  }
+
+  /// HTTP Protocol Version type
+  enum HTTPProtocolVersion {
+    case http1
+    case http2
+  }
+
+  public func channelRead(ctx: ChannelHandlerContext, data: NIOAny) {
+    // Detect the HTTP protocol version for the incoming request, or error out if it
+    // couldn't be detected.
+    var inBuffer = unwrapInboundIn(data)
+    guard let initialData = inBuffer.readString(length: inBuffer.readableBytes),
+          let preamble = initialData.split(separator: "\r\n",
+                                           maxSplits: 1,
+                                           omittingEmptySubsequences: true).first,
+          let version = protocolVersion(String(preamble)) else {
+
+      ctx.fireErrorCaught(HTTPProtocolVersionError.invalidHTTPProtocolVersion)
+      return
+    }
+
+    // Depending on whether it is HTTP1 or HTTP2, created different processing pipelines.
+    // Inbound handlers in handlersInitializer should expect HTTPServerRequestPart objects
+    // and outbound handlers should return HTTPServerResponsePart objects.
+    switch version {
+    case .http1:
+      // Upgrade connections are not handled since gRPC connections already arrive in HTTP2,
+      // while gRPC-Web does not support HTTP2 at all, so there are no compelling use cases
+      // to support this.
+      _ = ctx.pipeline.configureHTTPServerPipeline(withErrorHandling: true)
+        .then { ctx.pipeline.add(handler: GRPCWebCorsHandler()) }
```

Comment:
I assume we do not need to add the `CORS` handler to `HTTP2` requests because we always expect those to be from a "true" gRPC client and not via Javascript from within a browser? Does gRPC-Web guarantee to never send its requests via HTTP2 (in which case we might need CORS support on the HTTP2 path as well)?

Reply:
gRPC web exists because browsers don't support HTTP2 :) If/when that happens, we can add the CORS to the HTTP2 pipeline too. I can't think of any case where an HTTP2 client would implement CORS, since it's a security feature for browsers.